src/lib/draw/GeometryMode.svelte

```diff
@@ -38,88 +41,58 @@
 
       gjScheme.update((gj) => {
         let feature = gj.features.find((f) => f.id == currentlyEditing)!;
-        delete feature.properties.hide_while_editing;
-        return gj;
-      });
-    }
-
-    currentlyEditing = null;
-    currentlyEditingControls = null;
-    mapHover.set(null);
-  }
-
-  // Handle successful edits
-  routeTool.addEventListenerSuccessRoute((editedRoute) => {
-    if ($currentMode == thisMode) {
-      gjScheme.update((gj) => {
-        let feature = gj.features.find((f) => f.id == currentlyEditing)!;
-        // TODO Hack around https://github.com/acteng/atip/issues/142
         if (!feature) {
           window.alert(
             "You loaded another file or cleared everything while editing. Your changes were lost."
           );
           return gj;
         }
-        // Keep the ID and any properties. Just copy over stuff from routeSnapper.
-        // TODO We're depending on implementation details here and knowing what to copy...
-        feature.properties.length_meters = editedRoute.properties.length_meters;
-        feature.properties.waypoints = editedRoute.properties.waypoints;
-        delete feature.properties.hide_while_editing;
-        feature.geometry = editedRoute.geometry;
-        return gj;
-      });
 
-      // Stay in this mode
-      currentlyEditing = null;
-      currentlyEditingControls = null;
-    }
-  });
+        // Show the feature again
+        delete feature.properties.hide_while_editing;
 
-  routeTool.addEventListenerSuccessArea((editedArea) => {
-    if ($currentMode == thisMode) {
-      gjScheme.update((gj) => {
-        let feature = gj.features.find((f) => f.id == currentlyEditing)!;
-        if (!feature) {
-          window.alert(
-            "You loaded another file or cleared everything while editing. Your changes were lost."
-          );
-          return gj;
+        // If there are unsaved edits to the feature, copy them over. If the
+        // user explicitly canceled, then a failure callback would've run.
+        if (unsavedFeature) {
+          updateFeature(feature, unsavedFeature);
         }
-        // Keep the ID and any properties. Just copy over stuff from routeSnapper.
-        // TODO We're depending on implementation details here and knowing what to copy...
-        feature.properties.waypoints = editedArea.properties.waypoints;
-        delete feature.properties.hide_while_editing;
-        feature.geometry = editedArea.geometry;
         return gj;
       });
-
-      // Stay in this mode
-      currentlyEditing = null;
-      currentlyEditingControls = null;
     }
-  });
 
-  for (let tool of [pointTool, polygonTool]) {
-    tool.addEventListenerSuccess((feature) => {
+    stopEditing();
+    mapHover.set(null);
+  }
+
+  // Handle successful edits
+  for (let tool of [pointTool, polygonTool, routeTool]) {
+    tool.addEventListenerSuccess((editedFeature) => {
```

Comment:
It may be easier to review individual commits in order. I also did some refactoring here to merge the `routeTool` event handlers for areas vs routes into one. The added type-safety wasn't useful at all; it just meant we duplicated a bunch of handler code in here.